File: android/app/src/main/java/com/saratoga/Saratoga.kt
package com.saratoga

import android.content.Context
import com.cactus.CactusTokenCallback
import com.cactus.cactusComplete
import com.cactus.cactusDestroy
import com.cactus.cactusEmbed
import com.cactus.cactusInit
import com.cactus.cactusReset
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.security.MessageDigest
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import java.util.TimeZone

/**
 * Saratoga — on-device voice-driven clinical co-pilot.
 *
 * Per utterance: embed → per-τ top-k retrieval → if gate passes, fire task via E4B.
 * Tasks (MVP): τ1 differential, τ3 red flags, τ4 med reconciliation.
 * All on-device. FHIR-lite evidence log SHA-chained.
 */
class Saratoga(private val ctx: Context, private val weightsDir: File) {

    data class Chunk(val id: String, val tau: String, val text: String)
    data class Hit(val chunk: Chunk, val score: Float)
    data class TaskOutput(val tau: String, val card: String, val topHit: Hit?)
    data class Outcome(
        val tasksFired: List<TaskOutput>,
        val fhirBundle: String? = null
    )

    fun interface OnSection {
        fun update(tau: String, textSoFar: String)
    }

    fun interface OnQuick {
        fun fire(tau: String, chunkId: String, score: Float, preview: String)
    }

    private var embedHandle: Long = 0L
    private var reasonHandle: Long = 0L
    private val chunks = mutableListOf<Chunk>()
    private val vectors = mutableListOf<FloatArray>()
    private var prevHash = "GENESIS"
    private val logFile: File
    private val fhirFile: File

    // Per-τ gate thresholds (cosine).
    private val gateThresholds = mapOf(
        "tau1" to 0.42f,
        "tau3" to 0.48f,
        "tau4" to 0.48f,
    )

    private val taskTitles = mapOf(
        "tau1" to "DIFFERENTIAL + MUST-RULE-OUTS",
        "tau3" to "RED FLAG",
        "tau4" to "MEDICATION REC",
    )

    private val combinedSystem = """
        You are Saratoga, an on-device clinical co-pilot. A clinician is seeing a patient.
        Given the transcript and retrieved guideline chunks for up to three tasks, emit
        EXACTLY these three sections in this order. Use 'none' for any section with no
        relevant chunks.

        Formatting rules (strict):
        - Plain text only. NO LaTeX ($...$, \text, \ge, etc).
        - Numbers as plain digits or words (50, RR 58, SpO2 90%), no math mode.
        - Cite chunk ids in [brackets] at end of each line.
        - MUST-RULE-OUT: list only conditions the transcript actually raises suspicion for.
        - In ACTION, include specific drugs with doses when the chunk provides them.

        ## DDX
        1. <condition> — <key feature> [chunk_id]
        2. <condition> — <key feature> [chunk_id]
        3. <condition> — <key feature> [chunk_id]
        MUST-RULE-OUT:
        * <item> [chunk_id]
        * <item> [chunk_id]

        ## RED FLAG
        ALERT: <condition> [chunk_id]
        ACTION:
        * <specific drug + dose or specific step>
        * <specific drug + dose or specific step>
        * <specific drug + dose or specific step>

        ## MED REC
        MED FLAG: <drug(s)>
        RISK: <one line> [chunk_id]
        ACTION: <one line with specific dose or step>

        Under 200 words total. No preamble. No disclaimers. Stop after MED REC.
    """.trimIndent()

    init {
        val dir = File(ctx.filesDir, "evidence").apply { mkdirs() }
        val stamp = SimpleDateFormat("yyyyMMdd-HHmmss", Locale.US).format(Date())
        logFile = File(dir, "session-$stamp.jsonl")
        fhirFile = File(dir, "session-$stamp.fhir-queue.jsonl")
    }

    suspend fun load() = withContext(Dispatchers.IO) {
        val embedPath = File(weightsDir, "qwen3-embedding-0.6b").absolutePath
        val reasonPath = File(weightsDir, "gemma-4-e4b-it").absolutePath
        embedHandle = cactusInit(embedPath, null, false)
        reasonHandle = cactusInit(reasonPath, null, false)
        loadCorpus()
        buildIndex()
    }

    private fun loadCorpus() {
        chunks.clear()
        val files = listOf(
            "tau1_differential.md" to "tau1",
            "tau3_redflags.md" to "tau3",
            "tau4_medrec.md" to "tau4",
        )
        for ((f, tau) in files) {
            val text = ctx.assets.open("corpus/$f").bufferedReader().use { it.readText() }
            for (block in text.split("\n---")) {
                val b = block.trim()
                val m = Regex("^##\\s+(\\S+)", RegexOption.MULTILINE).find(b) ?: continue
                val id = m.groupValues[1].trim()
                val body = b.replaceFirst(Regex("^##\\s+\\S+\\s*\\n", RegexOption.MULTILINE), "").trim()
                if (body.isNotEmpty()) chunks.add(Chunk(id, tau, body))
            }
        }
    }

    private fun buildIndex() {
        vectors.clear()
        for (c in chunks) {
            val v = cactusEmbed(embedHandle, c.text, true)
            vectors.add(v)
        }
    }

    private fun ragTopPerTau(query: String, perTauK: Int = 3): Map<String, List<Hit>> {
        val q = cactusEmbed(embedHandle, query, true)
        val scored = chunks.indices.map { i ->
            val v = vectors[i]; var dot = 0f
            for (j in v.indices) dot += v[j] * q[j]
            Hit(chunks[i], dot)
        }
        return scored.groupBy { it.chunk.tau }
            .mapValues { (_, lst) -> lst.sortedByDescending { it.score }.take(perTauK) }
    }

    private fun composeAllStreaming(
        utterance: String,
        perTau: Map<String, List<Hit>>,
        firedTaus: Set<String>,
        onSection: OnSection?,
    ): Map<String, String> {
        val labels = mapOf(
            "tau1" to "Differential chunks",
            "tau3" to "Red-flag chunks",
            "tau4" to "Medication chunks",
        )
        val blocks = firedTaus.map { tau ->
            val hits = perTau[tau].orEmpty().take(2)
            val lines = hits.joinToString("\n") { "- [${it.chunk.id}] ${it.chunk.text.take(180)}" }
            "${labels[tau]}:\n$lines"
        }.joinToString("\n\n")
        val user = "Transcript:\n\"$utterance\"\n\n$blocks\n\nEmit the three sections."
        val messages = JSONArray()
            .put(JSONObject().put("role", "system").put("content", combinedSystem))
            .put(JSONObject().put("role", "user").put("content", user))
        val opts = JSONObject().put("max_tokens", 260).put("temperature", 0.2).toString()

        val buf = StringBuilder()
        // Strict header anchor to avoid accidental substring matches mid-stream.
        val headerRe = Regex("(?m)^##\\s+(DDX|RED FLAG|MED REC)\\b.*$")
        val cb = if (onSection != null) CactusTokenCallback { token, _ ->
            buf.append(token)
            val full = buf.toString()
            val matches = headerRe.findAll(full).toList()
            if (matches.isEmpty()) return@CactusTokenCallback
            // Update all firing tau panes with their respective slice (cleaner mid-stream).
            for ((i, m) in matches.withIndex()) {
                val t = headerToTau(m.groupValues[1]) ?: continue
                if (t !in firedTaus) continue
                val start = m.range.last + 1
                // Strip any partial trailing header fragment.
                var end = if (i + 1 < matches.size) matches[i + 1].range.first else full.length
                val tail = Regex("\\n##[^\\n]*$").find(full.substring(start, end))
                if (tail != null) end = start + tail.range.first
                val body = stripLatex(full.substring(start, end).trim())
                if (body.isNotEmpty()) onSection.update(t, body)
            }
        } else null

        cactusReset(reasonHandle)
        cactusComplete(reasonHandle, messages.toString(), opts, null, cb)
        return parseSections(buf.toString(), firedTaus)
    }

    private fun headerToTau(h: String): String? = when (h) {
        "DDX" -> "tau1"
        "RED FLAG" -> "tau3"
        "MED REC" -> "tau4"
        else -> null
    }

    private fun parseSections(text: String, firedTaus: Set<String>): Map<String, String> {
        val out = mutableMapOf<String, String>()
        val regex = Regex("(?m)^##\\s+(DDX|RED FLAG|MED REC)\\b.*$")
        val matches = regex.findAll(text).toList()
        for ((i, m) in matches.withIndex()) {
            val header = m.groupValues[1]
            val tau = when (header) {
                "DDX" -> "tau1"
                "RED FLAG" -> "tau3"
                "MED REC" -> "tau4"
                else -> continue
            }
            if (tau !in firedTaus) continue
            val start = m.range.last + 1
            val end = if (i + 1 < matches.size) matches[i + 1].range.first else text.length
            val body = stripLatex(text.substring(start, end).trim())
            if (body.isNotBlank() && !body.equals("none", ignoreCase = true)) out[tau] = body
        }
        return out
    }

    private fun stripLatex(s: String): String {
        // Remove \text{X}, $...$, \ge, \le, etc. Keep readable text.
        var t = s
        t = Regex("\\\\text\\{([^}]*)\\}").replace(t) { it.groupValues[1] }
        t = Regex("\\$([^$]*)\\$").replace(t) { it.groupValues[1] }
        t = t.replace("\\ge", "≥").replace("\\le", "≤")
            .replace("\\times", "×").replace("\\pm", "±")
            .replace("\\%", "%").replace("\\_", "_")
        t = Regex("\\\\[a-zA-Z]+\\b").replace(t, "")
        return t
    }

    suspend fun handle(
        utterance: String,
        onQuick: OnQuick? = null,
        onSection: OnSection? = null,
    ): Outcome = withContext(Dispatchers.Default) {
        appendLog("transcript", utterance)
        val perTau = ragTopPerTau(utterance)
        val firedTaus = perTau.entries
            .filter { it.value.isNotEmpty() && it.value[0].score >= (gateThresholds[it.key] ?: 0.5f) }
            .map { it.key }.toSet()
        // Stage 1 (sub-second): fire quick previews from RAG hits alone.
        for (tau in firedTaus) {
            val top = perTau[tau]!!.first()
            val preview = top.chunk.text.take(160)
            appendLog("quick:$tau", "${top.chunk.id} s=${"%.2f".format(top.score)}")
            onQuick?.fire(tau, top.chunk.id, top.score, preview)
        }
        val cards = if (firedTaus.isNotEmpty())
            composeAllStreaming(utterance, perTau, firedTaus, onSection)
        else emptyMap()
        val fired = firedTaus.mapNotNull { tau ->
            val card = cards[tau] ?: return@mapNotNull null
            val top = perTau[tau]?.firstOrNull()
            appendLog("task:$tau", card)
            TaskOutput(tau, card, top)
        }
        val bundle = if (fired.isNotEmpty()) emitFhirBundle(utterance, fired) else null
        Outcome(fired, bundle)
    }

    private fun emitFhirBundle(utterance: String, fired: List<TaskOutput>): String {
        val encId = "enc-${System.currentTimeMillis()}"
        val bundle = JSONObject()
            .put("resourceType", "Bundle")
            .put("type", "collection")
            .put("entry", JSONArray().apply {
                put(JSONObject().put("resource", JSONObject()
                    .put("resourceType", "Encounter")
                    .put("id", encId)
                    .put("status", "in-progress")
                    .put("class", "ambulatory")
                    .put("period", JSONObject()
                        .put("start", isoNow()))))
                put(JSONObject().put("resource", JSONObject()
                    .put("resourceType", "ClinicalImpression")
                    .put("encounter", JSONObject().put("reference", "Encounter/$encId"))
                    .put("description", utterance.take(400))
                    .put("finding", JSONArray().apply {
                        fired.forEach {
                            put(JSONObject()
                                .put("itemCodeableConcept", JSONObject()
                                    .put("text", "${it.tau}:${it.topHit?.chunk?.id ?: "?"}"))
                                .put("basis", it.card.take(600)))
                        }
                    })))
            })
        fhirFile.appendText(bundle.toString() + "\n")
        appendLog("fhir_queued", "enc=$encId")
        return encId
    }

    fun syncQueueSize(): Int = if (fhirFile.exists()) fhirFile.readLines().count { it.isNotBlank() } else 0

    fun drainSyncQueue(): Pair<Int, List<String>> {
        if (!fhirFile.exists()) return 0 to emptyList()
        val lines = fhirFile.readLines().filter { it.isNotBlank() }
        val bundles = mutableListOf<String>()
        lines.forEach { line ->
            try {
                val enc = JSONObject(line).getJSONArray("entry").getJSONObject(0)
                    .getJSONObject("resource").optString("id", "?")
                bundles.add(enc)
            } catch (_: Throwable) {
            }
        }
        fhirFile.delete()
        appendLog("sync_drained", "${lines.size}")
        return lines.size to bundles
    }

    private fun appendLog(kind: String, text: String) {
        val t = isoNow()
        val h = sha256(prevHash + t + text)
        val entry = JSONObject()
            .put("t_iso", t).put("kind", kind).put("text", text)
            .put("prev_hash", prevHash).put("hash", h)
        logFile.appendText(entry.toString() + "\n")
        prevHash = h
    }

    fun seal(): String {
        val manifest = JSONObject()
            .put("session_id", logFile.nameWithoutExtension)
            .put("tip_hash", prevHash)
            .put("sealed_at", isoNow())
        File(logFile.parentFile, "${logFile.nameWithoutExtension}.seal.json").writeText(manifest.toString(2))
        return prevHash
    }

    fun close() {
        if (reasonHandle != 0L) { cactusDestroy(reasonHandle); reasonHandle = 0L }
        if (embedHandle != 0L) { cactusDestroy(embedHandle); embedHandle = 0L }
    }

    fun taskTitle(tau: String): String = taskTitles[tau] ?: tau

    private fun isoNow(): String = SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'", Locale.US).apply {
        timeZone = TimeZone.getTimeZone("UTC")
    }.format(Date())

    private fun sha256(s: String): String =
        MessageDigest.getInstance("SHA-256").digest(s.toByteArray()).joinToString("") { "%02x".format(it) }
}

File: ios/SaratogaApp/Saratoga.swift
import Foundation
import cactus

/// Saratoga clinical co-pilot engine.
final class Saratoga {

    struct Chunk {
        let id: String
        let tau: String
        let text: String
    }
    struct Hit {
        let chunk: Chunk
        let score: Float
    }
    struct TaskOutput {
        let tau: String
        let card: String
        let topHit: Hit?
    }
    struct Outcome {
        let tasksFired: [TaskOutput]
        let fhirBundle: String?
    }

    private var embed: CactusModelT?
    private var reason: CactusModelT?
    private var chunks: [Chunk] = []
    private var vectors: [[Float]] = []
    private var prevHash = "GENESIS"
    private let logURL: URL
    private let fhirURL: URL

    private let gate: [String: Float] = ["tau1": 0.42, "tau3": 0.48, "tau4": 0.48]
    private let labels: [String: String] = [
        "tau1": "Differential chunks",
        "tau3": "Red-flag chunks",
        "tau4": "Medication chunks"
    ]

    private let combinedSystem = """
    You are Saratoga, an on-device clinical co-pilot. A clinician is seeing a patient.
    Given the transcript and retrieved guideline chunks for up to three tasks, emit
    EXACTLY these three sections in this order. Use 'none' for any section with no
    relevant chunks.

    Formatting rules (strict):
    - Plain text only. NO LaTeX.
    - Numbers as plain digits or words (50, RR 58, SpO2 90%).
    - Cite chunk ids in [brackets].
    - MUST-RULE-OUT: only conditions the transcript raises suspicion for.
    - In ACTION, include specific drugs with doses when the chunk provides them.

    ## DDX
    1. <condition> — <key feature> [chunk_id]
    2. <condition> — <key feature> [chunk_id]
    3. <condition> — <key feature> [chunk_id]
    MUST-RULE-OUT:
    * <item> [chunk_id]
    * <item> [chunk_id]

    ## RED FLAG
    ALERT: <condition> [chunk_id]
    ACTION:
    * <specific step>
    * <specific step>
    * <specific step>

    ## MED REC
    MED FLAG: <drug(s)>
    RISK: <one line> [chunk_id]
    ACTION: <one line>

    Under 200 words total. No preamble. Stop after MED REC.
    """

    init(documentsURL: URL) {
        let dir = documentsURL.appendingPathComponent("evidence")
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let fmt = DateFormatter()
        fmt.dateFormat = "yyyyMMdd-HHmmss"
        let stamp = fmt.string(from: Date())
        logURL = dir.appendingPathComponent("session-\(stamp).jsonl")
        fhirURL = dir.appendingPathComponent("session-\(stamp).fhir-queue.jsonl")
    }

    func load(weightsDir: URL) throws {
        // iPhone free Dev memory budget is tight (~3-5 GB RSS). Strategy:
        // 1. Load precomputed index from app bundle (no embed calls during init).
        // 2. Load embedder ONLY (small, ~500 MB) for runtime query embeds.
        // 3. DEFER reason model load to first handle() call.
        let fm = FileManager.default
        let e2b = weightsDir.appendingPathComponent("gemma-4-e2b-it")
        let e4b = weightsDir.appendingPathComponent("gemma-4-e4b-it")
        let reasonDir: URL = fm.fileExists(atPath: e2b.appendingPathComponent("config.txt").path)
            ? e2b : e4b
        Saratoga.unprotectTree(weightsDir.appendingPathComponent("qwen3-embedding-0.6b"))
        Saratoga.unprotectTree(reasonDir)
        self.reasonWeightsPath = reasonDir.path

        if !loadPrecomputedIndex() {
            // fallback: old path (will embed on device)
            let embedPath = weightsDir.appendingPathComponent("qwen3-embedding-0.6b").path
            embed = try cactusInit(embedPath, nil, false)
            loadCorpus()
            try buildIndex()
            return
        }
        // Normal path — tiny embedder for query-time only
        let embedPath = weightsDir.appendingPathComponent("qwen3-embedding-0.6b").path
        embed = try cactusInit(embedPath, nil, false)
    }

    private func loadPrecomputedIndex() -> Bool {
        guard let url = Bundle.main.url(forResource: "corpus_index",
                                         withExtension: "json") else { return false }
        guard let data = try? Data(contentsOf: url),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let chunkList = obj["chunks"] as? [[String: Any]],
              let vecList = obj["vectors"] as? [[Double]] else { return false }
        chunks.removeAll()
        vectors.removeAll()
        for (i, c) in chunkList.enumerated() {
            guard let id = c["id"] as? String,
                  let topicRaw = c["topic"] as? String,
                  let text = c["text"] as? String,
                  i < vecList.count else { continue }
            // Map corpus topic names to our τ keys.
            let tau: String
            if topicRaw.hasPrefix("tau1") { tau = "tau1" }
            else if topicRaw.hasPrefix("tau3") { tau = "tau3" }
            else if topicRaw.hasPrefix("tau4") { tau = "tau4" }
            else { continue }
            chunks.append(Chunk(id: id, tau: tau, text: text))
            vectors.append(vecList[i].map { Float($0) })
        }
        return !chunks.isEmpty
    }

    private var reasonWeightsPath: String = ""

    /// Lazy-open the reason model on first use.
    private func ensureReason() throws {
        if reason != nil { return }
        reason = try cactusInit(reasonWeightsPath, nil, false)
    }

    /// Recursively mark files as NSFileProtectionNone so mmap works while the
    /// device is unlocked and while the app is running (iOS encrypts Documents
    /// by default; mmap on Class A / B files can fail intermittently).
    static func unprotectTree(_ root: URL) {
        let fm = FileManager.default
        try? (root as NSURL).setResourceValue(URLFileProtection.none, forKey: .fileProtectionKey)
        guard let e = fm.enumerator(at: root,
                                    includingPropertiesForKeys: [.isDirectoryKey],
                                    options: []) else { return }
        for case let url as URL in e {
            try? (url as NSURL).setResourceValue(URLFileProtection.none, forKey: .fileProtectionKey)
        }
    }

    private func loadCorpus() {
        chunks.removeAll()
        let files: [(String, String)] = [
            ("tau1_differential", "tau1"),
            ("tau3_redflags", "tau3"),
            ("tau4_medrec", "tau4"),
        ]
        for (stem, tau) in files {
            guard let url = Bundle.main.url(forResource: stem, withExtension: "md",
                                             subdirectory: "Resources/corpus")
                ?? Bundle.main.url(forResource: stem, withExtension: "md")
                ?? Bundle.main.url(forResource: stem, withExtension: "md",
                                    subdirectory: "corpus"),
                  let text = try? String(contentsOf: url) else { continue }
            for block in text.components(separatedBy: "\n---") {
                let b = block.trimmingCharacters(in: .whitespacesAndNewlines)
                guard let m = b.range(of: "^##\\s+(\\S+)",
                                       options: [.regularExpression, .anchored]) else { continue }
                let header = String(b[m])
                let id = header.replacingOccurrences(
                    of: "^##\\s+", with: "",
                    options: .regularExpression
                ).trimmingCharacters(in: .whitespaces)
                let body = b.replacingOccurrences(
                    of: "^##\\s+\\S+\\s*\\n", with: "",
                    options: .regularExpression
                ).trimmingCharacters(in: .whitespacesAndNewlines)
                if !body.isEmpty { chunks.append(Chunk(id: id, tau: tau, text: body)) }
            }
        }
    }

    private func buildIndex() throws {
        guard let embed = embed else { return }
        vectors.removeAll()
        for c in chunks {
            let v = try cactusEmbed(embed, c.text, true)
            vectors.append(v)
        }
    }

    private func ragTopPerTau(_ query: String, k: Int = 3) throws -> [String: [Hit]] {
        guard let embed = embed else { return [:] }
        let q = try cactusEmbed(embed, query, true)
        let scored: [Hit] = zip(chunks, vectors).map { (chunk, v) in
            var dot: Float = 0
            for j in 0..<v.count { dot += v[j] * q[j] }
            return Hit(chunk: chunk, score: dot)
        }
        let grouped = Dictionary(grouping: scored) { $0.chunk.tau }
        return grouped.mapValues { $0.sorted { $0.score > $1.score }.prefix(k).map { $0 } }
    }

    private func composeAll(utterance: String,
                             perTau: [String: [Hit]],
                             fired: Set<String>,
                             onSection: ((String, String) -> Void)?) throws -> [String: String] {
        try ensureReason()
        guard let reason = reason else { return [:] }
        let blocks = fired.map { tau -> String in
            let hits = Array((perTau[tau] ?? []).prefix(2))
            let lines = hits.map { "- [\($0.chunk.id)] \(String($0.chunk.text.prefix(180)))" }
                .joined(separator: "\n")
            return "\(labels[tau] ?? tau):\n\(lines)"
        }.joined(separator: "\n\n")
        let user = "Transcript:\n\"\(utterance)\"\n\n\(blocks)\n\nEmit the three sections."
        let messages: [[String: String]] = [
            ["role": "system", "content": combinedSystem],
            ["role": "user", "content": user]
        ]
        let messagesJson = String(data: try JSONSerialization.data(withJSONObject: messages), encoding: .utf8)!
        let opts = "{\"max_tokens\":260,\"temperature\":0.2}"

        var buf = ""
        let headerRe = try NSRegularExpression(pattern: "(?m)^##\\s+(DDX|RED FLAG|MED REC)\\b.*$")
        let cb: ((String, UInt32) -> Void)? = onSection.map { callback in
            return { token, _ in
                buf += token
                let full = buf
                let matches = headerRe.matches(in: full, range: NSRange(full.startIndex..., in: full))
                if matches.isEmpty { return }
                for (i, m) in matches.enumerated() {
                    let nameRange = m.range(at: 1)
                    guard nameRange.location != NSNotFound,
                          let r = Range(nameRange, in: full) else { continue }
                    let header = String(full[r])
                    guard let t = ["DDX": "tau1", "RED FLAG": "tau3", "MED REC": "tau4"][header] else { continue }
                    if !fired.contains(t) { continue }
                    let start = full.index(full.startIndex, offsetBy: m.range.location + m.range.length)
                    let end = i + 1 < matches.count
                        ? full.index(full.startIndex, offsetBy: matches[i + 1].range.location)
                        : full.endIndex
                    var body = String(full[start..<end]).trimmingCharacters(in: .whitespacesAndNewlines)
                    body = Saratoga.stripLatex(body)
                    if !body.isEmpty { callback(t, body) }
                }
            }
        }

        cactusReset(reason)
        _ = try cactusComplete(reason, messagesJson, opts, nil, cb, nil)
        return parseSections(buf, fired: fired)
    }

    private func parseSections(_ text: String, fired: Set<String>) -> [String: String] {
        var out: [String: String] = [:]
        let headerRe = try! NSRegularExpression(pattern: "(?m)^##\\s+(DDX|RED FLAG|MED REC)\\b.*$")
        let matches = headerRe.matches(in: text, range: NSRange(text.startIndex..., in: text))
        for (i, m) in matches.enumerated() {
            guard let r = Range(m.range(at: 1), in: text) else { continue }
            let header = String(text[r])
            guard let tau = ["DDX": "tau1", "RED FLAG": "tau3", "MED REC": "tau4"][header] else { continue }
            if !fired.contains(tau) { continue }
            let start = text.index(text.startIndex, offsetBy: m.range.location + m.range.length)
            let end = i + 1 < matches.count
                ? text.index(text.startIndex, offsetBy: matches[i + 1].range.location)
                : text.endIndex
            var body = String(text[start..<end]).trimmingCharacters(in: .whitespacesAndNewlines)
            body = Saratoga.stripLatex(body)
            if !body.isEmpty && body.lowercased() != "none" { out[tau] = body }
        }
        return out
    }

    static func stripLatex(_ s: String) -> String {
        var t = s
        t = t.replacingOccurrences(
            of: "\\\\text\\{([^}]*)\\}", with: "$1", options: .regularExpression
        )
        t = t.replacingOccurrences(
            of: "\\$([^$]*)\\$", with: "$1", options: .regularExpression
        )
        t = t.replacingOccurrences(of: "\\ge", with: "≥")
            .replacingOccurrences(of: "\\le", with: "≤")
            .replacingOccurrences(of: "\\%", with: "%")
        t = t.replacingOccurrences(
            of: "\\\\[a-zA-Z]+\\b", with: "", options: .regularExpression
        )
        return t
    }

    func handle(utterance: String,
                onQuick: ((String, String, Float, String) -> Void)? = nil,
                onSection: ((String, String) -> Void)? = nil) throws -> Outcome {
        append(kind: "transcript", text: utterance)
        let perTau = try ragTopPerTau(utterance)
        let firedTaus: Set<String> = Set(perTau.compactMap { (tau, hits) -> String? in
            guard let top = hits.first else { return nil }
            return top.score >= (gate[tau] ?? 0.5) ? tau : nil
        })
        for tau in firedTaus {
            guard let top = perTau[tau]?.first else { continue }
            let preview = String(top.chunk.text.prefix(160))
            append(kind: "quick:\(tau)", text: "\(top.chunk.id) s=\(top.score)")
            onQuick?(tau, top.chunk.id, top.score, preview)
        }
        let cards = firedTaus.isEmpty
            ? [:]
            : try composeAll(utterance: utterance, perTau: perTau, fired: firedTaus, onSection: onSection)
        let fired: [TaskOutput] = firedTaus.compactMap { tau in
            guard let card = cards[tau] else { return nil }
            let top = perTau[tau]?.first
            append(kind: "task:\(tau)", text: card)
            return TaskOutput(tau: tau, card: card, topHit: top)
        }
        let bundle = fired.isEmpty ? nil : emitFhirBundle(utterance: utterance, fired: fired)
        return Outcome(tasksFired: fired, fhirBundle: bundle)
    }

    private func emitFhirBundle(utterance: String, fired: [TaskOutput]) -> String {
        let encId = "enc-\(Int(Date().timeIntervalSince1970 * 1000))"
        let bundle: [String: Any] = [
            "resourceType": "Bundle",
            "type": "collection",
            "entry": [
                ["resource": [
                    "resourceType": "Encounter",
                    "id": encId,
                    "status": "in-progress",
                    "class": "ambulatory",
                    "period": ["start": isoNow()]
                ]],
                ["resource": [
                    "resourceType": "ClinicalImpression",
                    "encounter": ["reference": "Encounter/\(encId)"],
                    "description": String(utterance.prefix(400)),
                    "finding": fired.map { t in
                        [
                            "itemCodeableConcept": [
                                "text": "\(t.tau):\(t.topHit?.chunk.id ?? "?")"
                            ],
                            "basis": String(t.card.prefix(600))
                        ]
                    }
                ]]
            ]
        ]
        if let data = try? JSONSerialization.data(withJSONObject: bundle, options: [.prettyPrinted]),
           let str = String(data: data, encoding: .utf8) {
            if let fh = try? FileHandle(forWritingTo: fhirURL) {
                fh.seekToEndOfFile(); fh.write(Data((str + "\n").utf8)); fh.closeFile()
            } else {
                try? (str + "\n").write(to: fhirURL, atomically: true, encoding: .utf8)
            }
        }
        append(kind: "fhir_queued", text: "enc=\(encId)")
        return encId
    }

    func syncQueueSize() -> Int {
        guard let txt = try? String(contentsOf: fhirURL) else { return 0 }
        return txt.split(whereSeparator: \.isNewline)
            .filter { !$0.trimmingCharacters(in: .whitespaces).isEmpty }
            .filter { $0.contains("resourceType") }
            .count > 0 ? 1 : 0
    }

    func drainSyncQueue() -> Int {
        guard FileManager.default.fileExists(atPath: fhirURL.path) else { return 0 }
        let n = syncQueueSize()
        try? FileManager.default.removeItem(at: fhirURL)
        append(kind: "sync_drained", text: "\(n)")
        return n
    }

    private func append(kind: String, text: String) {
        let t = isoNow()
        let h = sha256(prevHash + t + text)
        let entry: [String: Any] = [
            "t_iso": t, "kind": kind, "text": text,
            "prev_hash": prevHash, "hash": h
        ]
        if let data = try? JSONSerialization.data(withJSONObject: entry),
           let line = String(data: data, encoding: .utf8) {
            if let fh = try? FileHandle(forWritingTo: logURL) {
                fh.seekToEndOfFile(); fh.write(Data((line + "\n").utf8)); fh.closeFile()
            } else {
                try? (line + "\n").write(to: logURL, atomically: true, encoding: .utf8)
            }
        }
        prevHash = h
    }

    func seal() -> String {
        let manifest: [String: Any] = [
            "session_id": logURL.deletingPathExtension().lastPathComponent,
            "tip_hash": prevHash,
            "sealed_at": isoNow()
        ]
        if let data = try? JSONSerialization.data(withJSONObject: manifest, options: [.prettyPrinted]),
           let str = String(data: data, encoding: .utf8) {
            let sealURL = logURL.deletingPathExtension().appendingPathExtension("seal.json")
            try? str.write(to: sealURL, atomically: true, encoding: .utf8)
        }
        return prevHash
    }

    func close() {
        if let r = reason { cactusDestroy(r); reason = nil }
        if let e = embed { cactusDestroy(e); embed = nil }
    }

    private func isoNow() -> String {
        let fmt = ISO8601DateFormatter()
        fmt.formatOptions = [.withInternetDateTime, .withTimeZone]
        return fmt.string(from: Date())
    }

    private func sha256(_ s: String) -> String {
        // Simple SHA-256 via CommonCrypto.
        let data = s.data(using: .utf8) ?? Data()
        var hash = [UInt8](repeating: 0, count: 32)
        data.withUnsafeBytes { ptr in
            _ = Saratoga.ccSha256(ptr.baseAddress, CC_LONG(data.count), &hash)
        }
        return hash.map { String(format: "%02x", $0) }.joined()
    }

    private static func ccSha256(_ data: UnsafeRawPointer?, _ len: CC_LONG, _ out: UnsafeMutablePointer<UInt8>) -> UnsafeMutablePointer<UInt8> {
        return CC_SHA256(data, len, out)!
    }
}

// MARK: - CommonCrypto bridging
import CommonCrypto
